syft/execution/protocol.py

```diff
@@ -1,479 +1,560 @@
+        protocol = Protocol(
+            name=protocol_function.__name__,
+            include_state=self.include_state,
+            forward_func=protocol_function,
+            state_tensors=self.state_tensors,
+            id=sy.ID_PROVIDER.pop(),
+            owner=sy.local_worker,
+        )
+
+        # Build the protocol automatically
+        if self.args_shape:
+            args_ = PlaceHolder.create_placeholders(self.args_shape)
+            try:
+                protocol.build(*args_)
+            except TypeError as e:
+                raise ValueError(
+                    "Automatic build using @func2protocol failed!\nCheck that:\n"
+                    " - you have provided the correct number of shapes in args_shape\n"
+                    " - you have no simple numbers like int or float as args. If you do "
+                    "so, please consider using a tensor instead."
+                )
+        return protocol
+
+
 class Protocol(AbstractObject):
     """
-    A Protocol coordinates a sequence of Plans, deploys them on distant workers
-    and run them in a single pass.
+    A Protocol stores a sequence of torch actions, just like a function.
 
-    It's a high level object which contains the logic of a complex computation
-    distributed across several workers. The main feature of Protocol is the
-    ability to be sent / searched / fetched back between workers, and finally
-    deployed to identified workers. So a user can design a protocol, upload it
-    to a cloud worker, and any other workers will be able to search for it,
-    download it, and apply the computation program it contains on the workers
-    that it is connected to.
+    A Protocol is intended to store a sequence of torch actions, just like a function,
+    but it allows to send this sequence of actions to remote workers and to keep a
+    reference to it. This way, to compute remotely this sequence of actions on some remote
+    input referenced through pointers, instead of sending multiple messages you need now to send a
+    single message with the references of the protocol and the pointers.
 
+    All arguments are optional.
 
     Args:
-        plans: a list of pairs (worker, plan). "worker" can be either a real
-            worker or a worker id or a string to represent a fictive worker. This
-            last case can be used at creation to specify that two plans should be
-            owned (or not owned) by the same worker at deployment. "plan" can
-            either be a Plan or a PointerPlan.
-        id: the Protocol id
-        owner: the Protocol owner
-        tags: the Protocol tags (used for search)
-        description: the Protocol description
+        name: the name of the name
+        state: store the protocol tensors like model parameters
+        include_state: if true, implies that the protocol is a function, else a class. If true, the
+            state is re-integrated in the args to be accessed within the function
```

Comment:
We don't have to do it now, but I think we can jettison `include_state`, because `Protocols` based on classes don't make much sense (yet?), so all `Protocols` will be traced from functions for the foreseeable future.

Reply:
Hum, why not writing a class Protocol? They're just 2 syntaxes for the same things, am I wrong?